src/repr/src/scalar.rs

```diff
@@ -1228,6 +1228,65 @@ impl<'a> ScalarType {
         }
     }
 
+    /// Whether the current type is contained by the given type.
+    ///
+    /// This is a variant of `base_eq` where we allow the nullability information
+    /// of the fields within a record to be more restrictive in the current type.
+    pub fn is_covered_by(&self, other: &ScalarType) -> bool {
```

Comment:
I think that the contents of this method should just replace the contents of `base_eq`. I think the fact that we do exact comparison of record fields but we do the more forgiving comparison for list/map/array element types means that there was a bug in how we compare record fields.